images/builders/mlflow/run.sh

```diff
@@ -70,18 +72,42 @@ EOF
 
 fi
 
+# This section looks for any entrypoint arguments configured in the MLProject
+# file and tranforms them into workflow step input parameters. It does this by
+# generating two things:
+# - WORKFLOW_PARAMS - `ENV` entries that go into the the trainer image Dockerfile,
+# configuring default values for individual arguments
+# - MLFLOW_ARGS - `mlflow run` arguments to be passed at runtime that translate
+# workflow step input parameter values into mlflow entrypoint argument values
+WORKFLOW_PARAMS=""
+MLFLOW_ARGS=""
+if [ -f "MLproject" ]; then
+    WORKFLOW_PARAMS=$(yq e -j MLproject | \
+    jq -r ".entry_points.${FUSEML_MLFLOW_ENTRYPOINT}.parameters | to_entries | .[] |
+        \"ENV FUSEML_\" + (.key|ascii_upcase) + \"=\" + (.value.default|tostring)")
+    MLFLOW_ARGS=$(yq e -j MLproject | \
+    jq -r "[.entry_points.${FUSEML_MLFLOW_ENTRYPOINT}.parameters | to_entries | .[] |
+        \"-P \" + .key + \"=\$FUSEML_\" + (.key|ascii_upcase)] | join(\" \")")
+fi
+
 tag=$(get_tag)
 
-# destination is the task output, when using the internal FuseML registry we need to reference the repository
-# using the localhost address (see https://github.com/fuseml/fuseml/issues/65).
-destination="${FUSEML_REGISTRY:-127.0.0.1:30500}/${repository}:${tag}"
+destination="${FUSEML_PULL_REGISTRY:-${FUSEML_REGISTRY}}/${repository}:${tag}"
 
 if docker-ls tags ${repository} -r http://${registry} -j | jq -re ".tags | index(\"${tag}\")" &>/dev/null; then
     echo "${repository}:${tag} already exists, not building"
 else
     echo "${repository}:${tag} not found in ${registry}, building..."
     mkdir -p .fuseml
     cp -r ${MLFLOW_DOCKERFILE}/* .fuseml/
+    {
+        echo "ENV FUSEML_MLFLOW_ENTRYPOINT ${FUSEML_MLFLOW_ENTRYPOINT}"
+        echo "$WORKFLOW_PARAMS"
+        echo "ENV FUSEML_MLFLOW_ARGS '${MLFLOW_ARGS}'"
+    } >> .fuseml/Dockerfile
```

Comment:
Why adapting the Dockergile? Can't you just pass the arguments as `--build-arg`s?

Reply:
I think it's just a matter of preference. Both are equivalent. I think of this step as being a Dockerfile generator, and updating the Dockerfile makes more sense from that perspective.

Reply:
I think it's just a matter of preference. Both are equivalent. I think of this step as being a Dockerfile generator, and updating the Dockerfile makes more sense from that perspective.

Reply:
Well, up until this PR, this wasn't _Dockerfile generator_. And the problem with this approach is that user can hardly understand what the Dockerfile does because there's hidden generator that modifies it.
With ARGS already present in Dockerfile, it's understandable what is being done just by looking at the Dockerfile, without the need to read some bash script.

Reply:
> the problem with this approach is that user can hardly understand what the Dockerfile does because there's hidden generator that modifies it.
> With ARGS already present in Dockerfile, it's understandable what is being done just by looking at the Dockerfile, without the need to read some bash script.

I don't think this argument is entirely valid. If by "user", you mean someone who is looking at the workflow step logs, then the args are equally hidden. Only when you enable verbose mode can you see the container building command (and its arguments) in the logs. But then so can you see the generated Dockerfile.

There are other reasons I would like to avoid using args instead of appending env vars to the dockerfile:
1. if there is a large number of entrypoint arguments, you end up with a long and difficult to read (and compose, I might add) list of command line arguments for the container build command. Whereas, it's clearer when those args are added to the Dockerfile and printed as a whole.
2. the Dockerfile append approach is more consistent with what we're doing with the other container images (similar layout means users have something they can relate to)

Reply:
... there may be a third reason that just occurred to me: Dockerfile args and Dockerfile env vars are different things. The Dockerfile uses arguments during the container building process, but what I need is to pass environment variables to the container at runtime. I'm not even sure what you suggest can be accomplished with Dockerfile arguments.

Reply:
Based on our conversation and after a fair share of refactoring, we don't need to alter the  MLFlow trainer Dockerfile _nor_ supply additional build arguments. The new approach loads the entrypoints from the MLproject file directly in the trainer step and attempts to dynamically map their entrypoints to the workflow step parameters supplied to the trainer step.